src/spikeinterface/sortingcomponents/motion/dredge.py

```diff
@@ -0,0 +1,1469 @@
+            dt_bins = min(
+                abs(i - j), abs(i + batch_size - j), abs(i - j - batch_size)
+            )
+            if max_dt_bins and dt_bins > max_dt_bins:
+                continue
+            weights_ = weights
+            if masks is not None:
+                weights_ = masks.T[i : i + batch_size] * weights
+            corr = normxcorr1d(
+                raster_a[i : i + batch_size],
+                raster_b[j : j + batch_size],
+                weights=weights_,
+                xmasks=None if xmasks is None else xmasks.T[j : j + batch_size],
+                padding=disp,
+                normalized=normalized,
+                centered=centered,
+            )
+            max_corr, best_disp_inds = torch.max(corr, dim=2)
+            best_disp = possible_displacement[best_disp_inds.cpu()]
+            D[i : i + batch_size, j : j + batch_size] = best_disp.T
+            C[i : i + batch_size, j : j + batch_size] = max_corr.cpu().T
+
+    return D, C
+
+
+# TODO charlie sam : at the moment this is a duplicate with small differences see motion_estimate.py same function
+def normxcorr1d(
+    template,
+    x,
+    weights=None,
+    xmasks=None,
+    centered=True,
+    normalized=True,
+    padding="same",
+    conv_engine="torch",
+):
+    """normxcorr1d: Normalized cross-correlation, optionally weighted
+
+    The API is like torch's F.conv1d, except I have accidentally
+    changed the position of input/weights -- template acts like weights,
+    and x acts like input.
+
+    Returns the cross-correlation of `template` and `x` at spatial lags
+    determined by `mode`. Useful for estimating the location of `template`
+    within `x`.
+
+    This might not be the most efficient implementation -- ideas welcome.
+    It uses a direct convolutional translation of the formula
+        corr = (E[XY] - EX EY) / sqrt(var X * var Y)
+
+    This also supports weights! In that case, the usual adaptation of
+    the above formula is made to the weighted case -- and all of the
+    normalizations are done per block in the same way.
+
+    Arguments
+    ---------
+    template : tensor, shape (num_templates, length)
+        The reference template signal
+    x : tensor, 1d shape (length,) or 2d shape (num_inputs, length)
+        The signal in which to find `template`
+    weights : tensor, shape (length,)
+        Will use weighted means, variances, covariances if supplied.
+    centered : bool
+        If true, means will be subtracted (per weighted patch).
+    normalized : bool
+        If true, normalize by the variance (per weighted patch).
+    padding : int, optional
+        How far to look? if unset, we'll use half the length
+    conv_engine : string, one of "torch", "numpy"
+        What library to use for computing cross-correlations.
+        If numpy, falls back to the scipy correlate function.
+
+    Returns
+    -------
+    corr : tensor
+    """
+
+
+    if conv_engine == "torch":
```

Comment:
add `assert HAVE_TORCH` or similar

Reply:
We are not anymore making import at module level but import locally and so this HAVE_TORCH is dessapearing slowly.